### PR TITLE
Tpetra: Rework MPIAdvance AllToAll unit test to use `Kokkos::View`

### PR DESCRIPTION
**Pull request by cwpearson**
@trilinos/tpetra 

## Motivation
To allow Tpetra to use MPI Advance with GPUs

## Stakeholder Feedback
@mikethebos @bienz2

### CODE REVIEW

**Comment by cwpearson** [comment]
@csiefer2 is it reasonable to just use the Kokkos default device here, or should we somehow go through Tpetra's node system?

File: packages/tpetra/core/test/MPIAdvance/NeighborAllToAllW.cpp
Comment (cwpearson) [inline packages/tpetra/core/test/MPIAdvance/NeighborAllToAllW.cpp:1]: Some compilers don't like VLAs even if it's not actually variable

File: packages/tpetra/core/test/MPIAdvance/AllToAll.cpp [hunk 1]
```diff
@@ -104,8 +108,10 @@ void test_nothing(MPI_Comm comm, bool nullBufs, bool sameBufs,
 
 /*! \brief all ranks send and receive some
  */
+template <typename Device>
 void test_random(MPI_Comm comm, int seed, Teuchos::FancyOStream &out,
                  bool &success) {
+  static_assert(Kokkos::is_device_v<Device>, "");
```
[the remark this answers is not in the document]
Comment (cwpearson) [reply packages/tpetra/core/test/MPIAdvance/AllToAll.cpp:114]: Just making sure `Device` is actually a `Kokkos::Device`. When Trilinos requires c++20 we can use `concepts` instead!

**Comment by jhux2** [comment]
CUDA PR testing appears to be down, @sebrowne is looking into it.

File: packages/tpetra/core/test/MPIAdvance/AllToAll.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (cwpearson) [reply packages/tpetra/core/test/MPIAdvance/AllToAll.cpp:114]: @csiefer2 is it okay?

**Pull request merged**